pkg/vmprovider/providers/vsphere2/vmprovider_vm_utils.go

```diff
@@ -447,3 +448,105 @@ func GetAttachedDiskUUIDToPVC(
 
 	return diskUUIDToPVC, nil
 }
+
+// GetAdditionalResourcesForBackup returns a list of Kubernetes client objects
+// that are relevant for VM backup (e.g. bootstrap referenced resources).
+func GetAdditionalResourcesForBackup(
+	vmCtx context.VirtualMachineContextA2,
+	k8sClient ctrlclient.Client) ([]ctrlclient.Object, error) {
+	var objects []ctrlclient.Object
+	// Get bootstrap related objects from CloudInit or Sysprep (mutually exclusive).
+	if bootstrapSpec := vmCtx.VM.Spec.Bootstrap; bootstrapSpec != nil {
+		if v := bootstrapSpec.CloudInit; v != nil {
+			if cooked := v.CloudConfig; cooked != nil {
+				out, err := cloudinit.GetSecretResources(vmCtx, k8sClient, vmCtx.VM.Namespace, *cooked)
+				if err != nil {
+					return nil, err
+				}
+				objects = append(objects, out...)
+			} else if raw := v.RawCloudConfig; raw != nil {
+				obj, err := getSecretOrConfigMapObject(vmCtx, k8sClient, raw.Name, true)
+				if err != nil {
+					return nil, err
+				}
+				objects = append(objects, obj)
+			}
+		} else if v := bootstrapSpec.Sysprep; v != nil {
+			if cooked := v.Sysprep; cooked != nil {
+				out, err := sysprep.GetSecretResources(vmCtx, k8sClient, vmCtx.VM.Namespace, cooked)
+				if err != nil {
+					return nil, err
+				}
+				objects = append(objects, out...)
+			} else if raw := v.RawSysprep; raw != nil {
+				obj, err := getSecretOrConfigMapObject(vmCtx, k8sClient, raw.Name, true)
+				if err != nil {
+					return nil, err
+				}
+				objects = append(objects, obj)
+			}
+		}
+
+		// Get bootstrap related objects from vAppConfig (can be used alongside LinuxPrep/Sysprep).
+		if vApp := bootstrapSpec.VAppConfig; vApp != nil {
+			if cooked := vApp.Properties; cooked != nil {
+				for _, p := range vApp.Properties {
+					if from := p.Value.From; from != nil {
+						// vAppConfig Properties are backed by Secret resources only.
+						obj, err := getSecretOrConfigMapObject(vmCtx, k8sClient, from.Name, false)
```

Comment:
Are these later deduped? I would expect the common case is to use the same secret for the properties.

Reply:
No, all the additional resources are stored in the VM's EC during backup individually.

Reply:
I see your point now. It makes sense to dedup here considering all the keys in properties can come from the same secret as you pointed out. Let me update the code to handle this case.

Reply:
Updated this function and its test code to dedup.